### DISCUSSION TITLE
Writing predictions for training and validation dataloaders on fit end

### DISCUSSION THREAD

**Discussion by adamklie** [Lightning-AI/pytorch-lightning#13018] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
Is there a simple way to take advantage of current Lightning features to write predictions on the entire training and validation sets to separate files on `fit` end? Been trying to take advantage of some combination of the `on_fit_end` and `BasePredictionWriter` (which I am already using for writing using `predict` functionality). Worst case I will just write a separate script but wanted to see if there was something built in I could take advantage of.

Thanks!

**Comment by rohitgr7** [comment]
there is no built-in component for it and baseprediction writer callback is the best one you can configure and use.

**Reply by adamklie** [reply]
How would you configure this callback to write predictions on the entire training and val sets to separate files only once?

**Reply by rohitgr7** [reply]
if you are doing epoch based training then you can do something like:

```python
class MyPredicitonWriter(Callback):
    def __init__(self):
        self.train_preds = []
        self.val_preds = []
 
    def on_train_batch_end(self, trainer, pl_module, outputs, *args, **kwargs):
        if trainer.current_epoch == trainer.max_epochs - 1:
            self.train_preds.append(outputs)

    def on_train_epoch_end(self, trainer, pl_module):
        if trainer.current_epoch == trainer.max_epochs - 1:
            # save train_preds to file

    def on_validation_batch_end(self, trainer, pl_module, outputs, *args, **kwargs)
         if trainer.current_epoch == trainer.max_epochs - 1:
             self.val_preds.append(outputs)
   
    def on_validation_epoch_end(self, trainer, pl_module):
         if trainer.current_epoch == trainer.max_epochs - 1:
             # save val_preds to file


trainer = Trainer(..., callbacks=[MyPredicitonWriter()])
```

**Reply by adamklie** [reply]
Ok got it. Does that mean that the `training_step` and `val_step` functions in my `pl_module` need to return what I expect to output to the file. Where would the `outputs` arguments in the above functions be coming from?

**Reply by rohitgr7** [reply]
yes.. the stuff that you return from the training/validation_step hooks.

**Reply by adamklie** [reply]
Sorry for the naivete here, but how would that work for the `training step` hook? Would I have to override the default `FitLoop` to extract the loss from the `training step` return? Just trying to make sure I'm not missing anything.

**Reply by rohitgr7** [reply]
Sorry, didn't get you. Can you explain with some pseudo code?

**Comment by adamklie** [comment]
Sure. As far as I understand, the default loops called when running `fit` in `pl` expect the training step to return just the loss. Something like:

```python
# FitLoop
for epoch in range(max_epochs):

    # TrainingEpochLoop
    for batch_idx, batch in enumerate(train_dataloader):

        # TrainingBatchLoop
        for split_batch in tbptt_split(batch):

            # OptimizerLoop
            for optimizer_idx, opt in enumerate(optimizers):

                loss = lightning_module.training_step(batch, batch_idx, optimizer_idx)
```

But in order to write predictions for the training data using something similar to the `MyPredictionWriter` you outlined above, I would need the `lightning_module`'s `training_step` to return not only the loss, but also the predictions.  So the training_step would return the loss and predictions in some kind of data structure
```python
def training_step(self, batch, batch_idx):
        # Compute predictions and loss
        return loss, predictions
```

So I would have thought I would need override this loop in order to handle the fact that the `training_step` hook returns more than just the loss. Something like:
```python
# FitLoop
for epoch in range(max_epochs):

    # TrainingEpochLoop
    for batch_idx, batch in enumerate(train_dataloader):

        # TrainingBatchLoop
        for split_batch in tbptt_split(batch):

            # OptimizerLoop
            for optimizer_idx, opt in enumerate(optimizers):

                loss = lightning_module.training_step(batch, batch_idx, optimizer_idx)[0]  # Grab the first index of the returned values
```

Does that make sense?

**Reply by rohitgr7** [reply]
you can return a dict

```python
return {'loss': loss, 'predictions': prediction}
```
make sure you have a `'loss'` key to let Lightning know what the loss is to backdrop.

ref: https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html?highlight=training_step#training-step

**Reply by ndahiya3** [reply]
If the predictions are returned on every training_step then are they ALL saved in RAM or GPU? With a large medical image segmentation dataset, is there is any way to control when predictions are returned? And then after saving, how can that memory be cleared? Say, just save/return predictions only 10 times every epoch? Is there any existing functionality to do this in lightning?

**Reply by adamklie** [reply]
So just to be clear, if I return a dict with the `'loss'` key in it I don't need to do anything else when it comes to the training loop. That is handled for me?

**Reply by rohitgr7** [reply]
@ndahiya3 I think they will stay on GPU, not 100% sure, but you can check with a subsample by looking at the device of outputs. If they are on GPU you can move them to CPU inside the training step before returning it.

> Say, just save/return predictions only 10 times every epoch

do you mean return the prediction for only 10 batches per epoch?


@adamklie yes, if you return a `'loss'` key then it will be handled automatically for the backward-prop.

**Reply by ndahiya3** [reply]
@rohitgr7 Yeah, return predictions say for say only 10 batches per epoch to save on hard drive. I can't save all the predictions on either GPU or CPU as dataset is large.

**Reply by rohitgr7** [reply]
@ndahiya3 

you can just add conditions based on `batch_idx`.
```py
def training_step(self, batch, batch_idx):
    preds = ...
    if batch_idx < 10:
        return {'loss': loss, 'preds': preds}
    return {'loss': loss'}

def training_epoch_end(self, outputs):
    for out in outputs:
        if 'pred' in out:
            save(out['pred'])
```

**Reply by ndahiya3** [reply]
@rohitgr7 Thanks. I think I'll do something like you mentioned. I am getting following warning: ~python3.9/site-packages/pytorch_lightning/loops/optimization/closure.py:35: LightningDeprecationWarning: One of the returned values {'preds'} has a `grad_fn`. We will detach it automatically but this behaviour will change in v1.6. Please detach it manually: `return {'loss': ..., 'something': something.detach()}`

Should prediction be detached in the training_step() ? Loss.backward() is yet to be run at this stage, right? If I detach prediction at this stage will it not interfere with the backward pass running correctly? Detached prediction will disconnect the loss from the rest of the network, I believe.

**Reply by semaphore-egg** [reply]
@ndahiya3 Unless you have some advanced purpose, let us detach every tensor, except the `loss`, to be returned from `training_step`.
`detached_t=source_t.detach()` means creating a new tensor `detached_t` which shares inner data with `source_t`. I.e., `shallow copy`. 

The source tensor `source_t` is still in the computation graph and its gradients are recorded once further differentiable operations are applied.

The new tensor `detached_t` is not in the computational graph. Operations on `detached_t` do not change the computational graph, except that directly modifying data of `detach_t` means modifying data of `source_t`.

A safer way is to make a deep copy `detached_t=source_t.detach().clone()`. In this way, `detached_t` and `source_t` are totally unrelated.

**Reply by rohitgr7** [reply]
@ndahiya3 

> If I detach prediction at this stage will it not interfere with the backward pass running correctly?

it won't. You can safely detach it since `loss` is already connected to the graph.
